File: apps/api/src/modules/users/mappers/user.mapper.ts
import type { AuthenticatedUser, UserProfile, UserProfileStats, UserSummary } from '@gamescore/types';

import type { AccountUser, ProfileUser, PublicUser } from '../repositories/user.repository';

/**
 * Mappers are the boundary between database rows and the HTTP contract. Going
 * through them is what guarantees a password hash or an email address cannot
 * leak into a public response by accident.
 */
export function toUserSummary(user: PublicUser): UserSummary {
  return {
    id: user.id,
    username: user.username,
    displayName: user.displayName,
    avatarUrl: user.avatarUrl,
    reputationScore: user.reputationScore,
    role: user.role,
    deleted: user.deletedAt !== null,
  };
}

export function toAuthenticatedUser(user: AccountUser): AuthenticatedUser {
  return {
    id: user.id,
    username: user.username,
    displayName: user.displayName,
    avatarUrl: user.avatarUrl,
    reputationScore: user.reputationScore,
    role: user.role,
    deleted: false,
    email: user.email,
    status: user.status,
    createdAt: user.createdAt.toISOString(),
    emailVerified: user.emailVerifiedAt !== null,
  };
}

export function toUserProfile(user: ProfileUser, stats: UserProfileStats): UserProfile {
  return {
    ...toUserSummary(user),
    bio: user.bio,
    createdAt: user.createdAt.toISOString(),
    stats,
  };
}

File: apps/api/src/GameScore.Infrastructure/Mappers/UserMapper.cs
using GameScore.Infrastructure.Persistence;

namespace GameScore.Infrastructure.Mappers;

public static class UserMapper
{
    public static object ToSummary(UserEntity user) =>
        new
        {
            id = user.Id,
            username = user.Username,
            displayName = user.DisplayName,
            avatarUrl = user.AvatarUrl,
            reputationScore = user.ReputationScore,
            role = user.Role.ToString(),
            deleted = user.DeletedAt is not null,
        };

    public static object ToAuthenticatedUser(UserEntity user) =>
        new
        {
            id = user.Id,
            username = user.Username,
            displayName = user.DisplayName,
            avatarUrl = user.AvatarUrl,
            reputationScore = user.ReputationScore,
            role = user.Role.ToString(),
            deleted = false,
            email = user.Email,
            status = user.Status.ToString(),
            createdAt = user.CreatedAt.ToString("O"),
            emailVerified = user.EmailVerifiedAt is not null,
        };

    public static object ToPlatform(PlatformEntity platform) =>
        new
        {
            id = platform.Id,
            slug = platform.Slug,
            name = platform.Name,
            abbreviation = platform.Abbreviation,
            family = platform.Family.ToString(),
        };
}
